linker/Linker.Steps/MarkStep.cs

```diff
@@ -217,6 +220,35 @@ void ProcessVirtualMethods ()
 			}
 		}
 
+		void ProcessActivatorCreateInstanceType ()
+		{
+			var markedTypes = Annotations.GetMarkedTypes ();
+			foreach (var information in _activatorCreateInstanceTypes) {
```

Comment:
Where is _activatorCreateInstanceTypes cleared?

Reply:
It's not and it's can't be.

Once an `Activator.CreateInstance` usage is registered in `_activatorCreateInstanceTypes` we need to reprocess the marking that usage triggers on each pass of the primary queue.  There are a number of queue phases that could cause new types to be marked.  If a new type is marked and meets the requirements of any of the `Activator.CreateInstance` usages then we'll need to mark that types ctor's accordingly.

The following happen after `ProcessActivatorCreateInstanceType` and could easily lead to new types being marked.
```
ProcessVirtualMethods ();
ProcessMarkedTypesWithInterfaces ();
DoAdditionalProcessing ();
```

and in more obscure cases `ProcessLazyAttributes()` and `ProcessLateMarkedAttributes()` could probably trigger a new type to be marked.

Reply:
I don't like this approach where we could iterate all the types over and over. If I understood the code this is handling scenario where we don't know which ctor to mark exactly.

Why don't we, in this case, build the dictionary of such types and process them at the end of `while (!QueueIsEmpty ()) {` and mark all their unmarked ctors then we clean the dictionary and be done for the pass.

Reply:
I don't think you are understanding why it's working this way.

Once we detect an `Activator.CreateInstance` usage we need to keep track of that for the entire duration of the MarkStep.  Think of each `Activator.CreateInstance` usage as creating a new semi-dynamic mark rule that the MarkStep needs to apply to all currently marked types and to all types that are marked later in the MrakStep.  We cannot clear this list of "dynamic mark logic".

Also when we detect an `Activator.CreateInstance` usage we need to retroactively apply the marking logic to all types that have already been marked before we came across this usage.  So I don't see a way to avoid the iterate all types process unless we build up a look up table during the TypeMapStep.  Building such a table seems costly in it's own right and you'd force every run to pay that price even if that set of assemblies makes no use of `Activator.CreateInstance`

What I could maybe do is 

* Process all marked types at the time the `Activator.CreateInstance` usage is detected.
* Then iterate `_activatorCreateInstanceTypes` during `MarkType`

Then we only pay the iterate all marked types cost once for each usage.  Would you prefer that?  Seems to me whether or not it will be less costly than the current approach depends on how usages are detected.  The current approach can batch multiple usages into one pass of all marked types.